Lombiq.Tests.UI/MonkeyTesting/MonkeyTester.cs

```diff
@@ -0,0 +1,213 @@
+- {nameof(MonkeyTestingOptions.GremlinsMogwais)}={string.Join(", ", _options.GremlinsMogwais)}
+- {nameof(MonkeyTestingOptions.GremlinsAttackDelay)}={_options.GremlinsAttackDelay.ToShortIntervalString()}");
+
+        private bool CanTestPage(PageMonkeyTestInfo pageTestInfo)
+        {
+            bool canTest = pageTestInfo.HasTimeToTest && ShouldTestPageUrl(pageTestInfo.Url);
+
+            if (!canTest)
+            {
+                Log.Info(
+                    !pageTestInfo.HasTimeToTest
+                    ? $"Available monkey testing time for \"{pageTestInfo.SanitizedUrl}\" is up and thus testing is complete."
+                    : $"Navigated to \"{pageTestInfo.Url}\" that should not be tested.");
+            }
+
+            return canTest;
+        }
+
+        private bool ShouldTestPageUrl(Uri url) => _options.UrlFilters.All(filter => filter.AllowUrl(_context, url));
+
+        private bool TryGetAvailablePageToTest(out PageMonkeyTestInfo pageTestInfo)
+        {
+            pageTestInfo = _visitedPages.FirstOrDefault(pageInfo => pageInfo.HasTimeToTest);
+            return pageTestInfo != null;
+        }
+
+        private PageMonkeyTestInfo GetCurrentPageTestInfo()
+        {
+            string urlAsString = _context.Driver.Url;
+            var url = new Uri(urlAsString);
+
+            var cleanUrl = CleanUrl(url);
+
+            var pageTestInfo = _visitedPages.FirstOrDefault(pageInfo => pageInfo.SanitizedUrl == cleanUrl)
+                ?? new PageMonkeyTestInfo(url, cleanUrl, _options.PageTestTime);
+
+            Log.Info($"Current page is \"{pageTestInfo.SanitizedUrl}\".");
+
+            return pageTestInfo;
+        }
+
+        private Uri CleanUrl(Uri url)
+        {
+            foreach (var cleaner in _options.UrlSanitizers) url = cleaner.Clean(_context, url);
+
+            return url;
+        }
+
+        private Task TestCurrentPageAsync(PageMonkeyTestInfo pageTestInfo)
+        {
+            int randomSeed = GetRandomSeed();
+
+            return TestCurrentPageWithRandomSeedAsync(pageTestInfo, randomSeed);
+        }
+
+        private Task TestCurrentPageWithRandomSeedAsync(PageMonkeyTestInfo pageTestInfo, int randomSeed)
+        {
+            Log.ExecuteSection(
+                new LogSection(
+#pragma warning disable S103 // Lines should not be too long
+                    $"Monkey test \"{pageTestInfo.SanitizedUrl}\" within {pageTestInfo.TimeToTest.ToShortIntervalString()} with {randomSeed} random seed."),
+#pragma warning restore S103 // Lines should not be too long
+                () =>
+                {
+                    var pageTestTimeLeft = TestCurrentPageAndMeasureTestTimeLeft(pageTestInfo.TimeToTest, randomSeed);
+                    pageTestInfo.TimeToTest = pageTestTimeLeft;
+                    if (!_visitedPages.Contains(pageTestInfo)) _visitedPages.Add(pageTestInfo);
+                });
+
+            return _context.TriggerAfterPageChangeEventAsync();
```

Comment:
Control can arrive here before Gremlin finishes. I've seen this only on the frontend though (with 60s Gremlin time) but probably that's only the reason because accessibility checks happen just on the frontend (if turned on) and accessibility failing on the Gremlin overlay surfaced this.

![image](https://user-images.githubusercontent.com/1976647/146620176-94a061b4-c980-4810-942f-fb75a537df33.png)

Reply:
I updated monkey testing functionality to stop Gremlins script explicitly. It should solve that issue,

Reply:
This is still happening. I've added `TestEmptyPageAsMonkey` that demonstrates the issue and also fails because accessibility checking trips on the gremlin markers. Please delete it and `TestContensAdminAsMonkeyRecursivelyShouldWorkWithAdminUser` once you're done.

![image](https://user-images.githubusercontent.com/1976647/149045116-7dd93426-e764-44c6-ad03-25f1b2c24204.png)

If you run the test, you'll get an accessibility report like this (it's HTML but GitHub doesn't let you upload that).
[AccessibilityReport.txt](https://github.com/Lombiq/UI-Testing-Toolbox/files/7850983/AccessibilityReport.txt)

Reply:
I've fixed that by cleaning Gremlins visual indicator elements. Check it out. Works for me fine.

Reply:
Please push to the OSOCE repo so the build can run.

Reply:
Pushed. Let's wait and see.

Reply:
While `TestEmptyPageAsMonkey` passes, Gremlin interactions still go on when the even is triggered:
![image](https://user-images.githubusercontent.com/1976647/150699846-9184f2e2-a4f4-4d4f-bf86-deefff4d1f07.png)

So if the timing happens to be different, the same issue will arise with accessibility checking, and for other event handlers it's otherwise an issue as well.

Reply:
Fixed that in another way by waiting for Gremlins indicators to disappear. This seems like the only 100% working solution.

Reply:
Agree, my mistake. I've updated Gremlins indicator selector.

Reply:
Looks good.